source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -43,15 +43,22 @@ void Filter::initiateCall(const Http::RequestHeaderMap& headers,
     context_extensions = maybe_merged_per_route_config.value().takeContextExtensions();
   }
 
-  // If metadata_context_namespaces is specified, pass matching metadata to the ext_authz service.
+  // If metadata_context_namespaces is specified, pass matching filter metadata and
+  // typed filter metadata to the ext_authz service.
   envoy::config::core::v3::Metadata metadata_context;
   const auto& request_metadata =
       decoder_callbacks_->streamInfo().dynamicMetadata().filter_metadata();
+  const auto& request_typed_metadata =
+      decoder_callbacks_->streamInfo().dynamicMetadata().typed_filter_metadata();
   for (const auto& context_key : config_->metadataContextNamespaces()) {
-    const auto& metadata_it = request_metadata.find(context_key);
-    if (metadata_it != request_metadata.end()) {
+    if (const auto& metadata_it = request_metadata.find(context_key);
+        metadata_it != request_metadata.end()) {
       (*metadata_context.mutable_filter_metadata())[metadata_it->first] = metadata_it->second;
     }
+    if (const auto& metadata_it = request_typed_metadata.find(context_key);
+        metadata_it != request_typed_metadata.end()) {
+      (*metadata_context.mutable_typed_filter_metadata())[metadata_it->first] = metadata_it->second;
```

Comment:
Looking at the definition of [Metadata](https://github.com/envoyproxy/envoy/blob/e898ab07843a0b84b6a19df3e320e93139dc6d35/api/envoy/config/core/v3/base.proto#L235) I'm thinking whether the metadata should be duplicated in both typed/non-typed fields or just passed in one of them.
Can you explain the reason you prefer the typed variant over the non-typed variant?
/wait-any

Reply:
I'm not sure that duplicating the same metadata over is the right thing to do at the moment.
One thing we can do is configure what field to use to send the data.
I'd also suggest reading [this blog post](https://blog.envoyproxy.io/dynamic-extensibility-and-protocol-buffers-dcd0bf0b8801), that may be relevant here.

Reply:
I don't understand what you meant here by "duplicating the same metadata", the change doesn't "duplicate" anything and it's not the "same" data, this change just allows us to use the now preferred "Any struct" mentioned in the blog you linked above (section "Update (2020–06–24)").

Honestly I don't think adding a field to decide whether we use typed or non-typed metadata helps a lot since this only applies to envoy and the external auth service still need to figure out which to extract based on an implicit contract. That being said, I can make that change if this is the only way we can merge it in.

Reply:
I think I didn't read the code correctly on my second pass, and was under the impression that the same {key->value} will _always_ be added to both filter_metadata and typed_filter_metadata.
Upon re-reading the code, I see now that it only adds the key->value according to the metadata it reads from.

I'm trying to see where items are added to the typed_filter_metadata, and if I haven't missed something, then it is done [here](https://github.com/envoyproxy/envoy/blob/26329cc6e3e87e061a37b08cc79f63ad257bd837/source/common/http/async_client_impl.cc#L88). All other places use the filer_metadata.

As you said, the design allows a server to send both of them in order to be backwards compatible with the two types of clients (as a server doesn't know a-priori which clients versions it is going to serve). That said, IIUC Envoy will only parse the input from the typed metadata if exists.
Do you see a case where a client needs to connect to a server without knowing what the server supports?

Reply:
Thanks Adi! As we discussed on slack, it's better to give client (envoy in this case) more control over what is sent in which way (typed or non-typed).

I’ve added a typed_metadata_context_namespaces option to allow user to have more control over which namespace should use typed metadata and which namespace should use the non-type ones. With this, users can still have both typed and non-typed metadata with the same key passed to the external auth service if they really want in order to help with the migration but this is totally the users' own choice. The rule of thumb for the auth service side is that they should always prefer typed metadata if such a key exists, however, there’s nothing preventing it from doing the opposite to make sure the migration requirements can be met.

Please take a look and let me know your thoughts, thank you.